File: 0044 Wildcard Matching/solution.py
#! python3

class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        m, n = len(s) + 1, len(p) + 1
        dp = [[False] * n for _ in range(2)]
        dp[0][0] = True
        
        for j in range(1, n):
            if p[j-1] == '*':
                dp[0][j] = dp[0][j-1]
                
        col = 0; #前一行
        for i in range(1, m):
            dp[col^1] = [False] * n
            for j in range(1, n):
                if p[j-1] == '?' or p[j-1] == s[i-1]:
                    dp[col^1][j] = dp[col][j-1]
                elif p[j-1] == '*':
                    dp[col^1][j] = dp[col][j] or dp[col][j-1] or dp[col^1][j-1]
            col ^= 1
        
        return dp[col][n-1]

if __name__ == "__main__":
    print(Solution().isMatch("cb", "?a"))
    print(Solution().isMatch("adceb", "*a*b"))

File: 0044 Wildcard Matching/solution.cpp
#include <bits/stdc++.h>
using std::string; using std::vector;

class Solution {
public:
    bool isMatch(string s, string p) {
        int n = s.size(), m = p.size();
        vector<vector<bool>> dp(n + 1, vector<bool> (m + 1, false));
        dp[0][0] = true;
        
        for (int i = 1; i <= m; i ++)
            if (p[i - 1] == '*' && dp[0][i - 1])
                dp[0][i] = true;
            else
                break;
        
        for (int i = 1; i <= n; i ++)
            for (int j = 1; j <= m; j ++)
            {
                if (s[i - 1] == p[j - 1] || p[j - 1] == '?')
                    dp[i][j] = dp[i - 1][j - 1];
                else if (p[j - 1] == '*')
                    dp[i][j] = dp[i - 1][j] | dp[i][j - 1];
            }
        return dp[n][m];
    }
};

int main(){
    Solution s;
    assert(s.isMatch("adceb", "*a*b") == true);
    assert(s.isMatch("cb", "?a") == false);
    return 0;
}